### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive when using the iteration variable to access object fields with different types

### ISSUE THREAD

**Issue by lantw44** [typescript-eslint/typescript-eslint#12556]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=C4TwDgpgBAYglhANgEwM5QLxQN4CgpQAWSiA9gPwBcUAdgK4C2ARhAE4Dc%2BUA7qaylShNSpRBACGNTgF9cuMcCgAzBClTV4SNJhzTOuJXygAKAMakaqRQGsIIKKSVQA2gHJiiMq4A0UV735kVwBdKHF0c0tgAEocLhUtVGdbEFDyciw6GmQIFRoIZBkgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Fields = {
  hello?: number;
  world?: boolean;
}

let fields: Fields = {};

for (const key of ['hello', 'world'] as const) {
  fields[key] ??= undefined;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error. Replacing the `['hello', 'world']` array with either `['hello']` or `['world']` produces no error.

### Actual Result

The `fields[key] ??= undefined` line causes the error:
Unnecessary conditional, left-hand side of `??` operator is always `null` or `undefined`.

**Comment by StyleShit** [comment]
Yeah, definitely looks like a bug

Thanks for reporting!

___

Interestingly, when both keys have the same type, there is no error:

```TS
type Fields = {
  hello?: number;
  world?: number;
}
```

**Comment by lantw44** [comment]
Yes, there is no error when two fields have at least one common type.
```typescript
type Fields = {
  hello?: boolean | number;
  world?: boolean;
}
```
```typescript
type Fields = {
  hello?: number;
  world?: 1 | 2 | 3;
}
```
```typescript
type Fields = {
  hello?: 1 | 2 | 4;
  world?: 1 | 3 | 5;
}
```

**Comment by gonzoblasco** [comment]
Interesting bug. The false positive only happens when the object fields have **different** types (`number | undefined` vs `boolean | undefined`), but disappears when they share a common type. That suggests the narrowing logic is collapsing the union to `never` somewhere in the iteration.

The playground link confirms it — `fields[key]` when `key` is `"hello" | "world"` should produce `number | boolean | undefined`, but the rule is reporting it as "always null or undefined", which means the type resolver is losing the `number | boolean` part and only seeing `undefined`.

Since the maintainers have labeled this `accepting prs`, I might take a look at the narrowing logic in `no-unnecessary-condition` if nobody else picks it up first.
